---
title: "Introduction to fastCNV - Visium HD data" 
output: 
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{Introduction to fastCNV - Visium HD data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


## Context 

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE, message = FALSE, warning = FALSE,
  comment = "#>"
)
```

`fastCNV` is a package to detect the putative Copy Number Variations (CNVs) in single cell (scRNAseq) data or Spatial Transcriptomics (ST) data. It also plots the computed CNVs.

In this vignette you will learn how to run this package on Visium HD seurat object. fastCNV runs on **Seurat5**. 

**Warning** : Visium HD samples are very RAM demanding and the session may crash if not enough RAM is available. We recommend running this vignette in a server. 

## Load packages


```{r install, eval=FALSE}
remotes::install_github("must-bioinfo/fastCNV")
```


```{r load}
library(fastCNV)
library(Seurat)
```

## Load the data

Load your Visium HD data with Seurat.

``` {r imports, eval = FALSE}

HDobj <- Load10X_Spatial(data.dir = "/path/to/your/data/", bin.size = c(8,16))

```

## Check annotation

You can load a separate annotation file for your dataset. It's usefull to have annotated spots for the analysis. 

```{r annots, eval=FALSE}
# Import the annotation file corresponding to your sample. For 10X ST data you can annotate your spots with LoupeBrowser
annotation_file <- read.csv("/path/to/your/annotations/for/VisiumHD.csv")

HD_obj[["annots"]] <- annotation_file$Annotations
```

You can take a look at which annotations are present in your objects, in order to select the labels of the cells that will be used as reference. 

```{r see_annots, output=FALSE, eval=FALSE, results='hide'}
unique(HD_obj[["annots"]])
```

## Run fastCNV 

During this step, fastCNV_10XHD() will use: 

- `CNVanalysis` : computes the CNV. If you have annotations for your Seurat object, you can add the parameters `referenceVar` and `referenceLabel`. 

- `CNVPerChromosome` : to compute the CNV per chromosome arm, and store it in the metadata of the Seurat object. This part of `fastCNV()` can be skipped by turning `getCNVPerChromosome` to `FALSE`.    

- `CNVcluster` : OPTIONAL - performs hierarchical clustering on a genomic score matrix extracted from a Seurat object. This may crash with large Visium HD samples.

- `plotCNVResultsHD` : to visualize the results (stored in the assays slot of the Seurat object).
This function will also build a PNG or PDF file for each sample containing their corresponding CNV heatmap in the current working directory, which can be changed using the `savePath` parameter.


We are going to run `fastCNV()` on our Visium HD object. The reference should be the non-tumor spots, for example, `Epithelium` or `Muscle`. 

```{r run1, eval = FALSE}
HD_obj <- fastCNV(HD_obj, sampleName = "HD_obj", referenceVar = "annots",
                      referenceLabel = c("Epithelium","Muscle"))
```


## CNV fraction

fastCNV also computes a `cnv_fraction` for each observation in the Seurat object. This can be directly plotted using Seurat spatial plotting functions, as we'll demonstrate on STColon1:

```{r plot_cnv_fraction, eval=FALSE}
library(Seurat)
library(patchwork)
SpatialFeaturePlot(HD_obj, "cnv_fraction") | SpatialPlot(HD_obj, group.by = "annots")
```


Here, we see some zones with higher CNV fractions than others. We can directly plot and test this:

```{r plot_cnv_fraction2, eval = FALSE}
library(ggplot2)
ggplot(FetchData(HD_obj, vars = c("annots", "cnv_fraction")), 
       aes(annots, cnv_fraction, fill = annots)) +
  geom_boxplot() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, color = "black"))
```


We can also plot the CNV per chromosome arm with the plotting functions from Seurat.

```{r plot_cnv_per_chromosome, eval=FALSE}
library(scales)

SpatialFeaturePlot(HD_obj, features = "11.q_CNV", pt.size.factor = 3)  +
  scale_fill_distiller(palette = "RdBu", direction = -1, limits = c(-1, 1), 
                       rescaler = function(x, to = c(0, 1), from = NULL) {
                         rescale_mid(x, to = to, mid = 0)
                       }) |
SpatialFeaturePlot(HD_obj, features = "6.q_CNV", pt.size.factor = 3) +
  scale_fill_distiller(palette = "RdBu", direction = -1, limits = c(-1, 1), 
                       rescaler = function(x, to = c(0, 1), from = NULL) {
                         rescale_mid(x, to = to, mid = 0)
                       })

```



## CNV classification

Using the CNV per chromosome arm and `CNVclassification()` we can get the alterations per chromosome arm (gain, loss or no alteration).


``` {r cnv_classification, eval=FALSE}
HD_obj <- CNVclassification(HD_obj)

common_theme <- theme(
  plot.title = element_text(size = 10),
  legend.text = element_text(size = 8),
  legend.title = element_text(size = 8),
  axis.title = element_text(size = 8),
  axis.text = element_text(size = 6)
)


SpatialDimPlot(HD_obj, group.by = "11.p_CNV_classification") +
  scale_fill_manual(values = c(gain = "red", no_alteration = "grey", loss = "blue")) +
  common_theme | SpatialDimPlot(HD_obj, group.by = "6.q_CNV_classification") +
  scale_fill_manual(values = c(gain = "red", no_alteration = "grey", loss = "blue")) +
  common_theme

```

## CNV clusters

In addition to the CNV fraction, fastCNV can also compute the CNV clusters. These can be plotted with the Seurat plotting functions.

```{r plot_cnv_clusters, eval=FALSE}
SpatialDimPlot(HD_obj, group.by = "cnv_clusters", pt.size.factor = 3) 
```


```{r cnv_props, eval=FALSE}
library(ggplot2)
library(SeuratObject)

ggplot(FetchData(HD_obj, vars = c("cnv_clusters", "annots")), aes(annots, fill = cnv_clusters)) +
  geom_bar(position = "fill") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, color = "black"))
```


## CNV tree

In this part of the vignette, we will show how to plot the CNV subclonality tree. CNV clusters are needed. 

First, we will build a metacell for each cluster, using `generateCNVClonesMatrix()`.  

```{r metacell_cluster, eval=FALSE}
cnv_matrix_clusters <- generateCNVClonesMatrix(HD_obj, healthyClusters = "1")

```

Now, we build the tree, using `CNVtree()`, `annotateCNVtree()` and `plotCNVtree()`.

  - `CNVtree()` will build the subclonality tree.
  
  - `annotateCNVtree()` will annotate each cnv gain/loss in the tree. 
  
  - `plotCNVtree()` will output the tree. 

```{r tree_cnv, eval=FALSE}
tree <- CNVtree(cnv_matrix_clusters)
tree_data <- annotateCNVtree(tree, cnv_matrix_clusters, 0.13)
plotCNVtree(tree_data)

```
