---
title: "Introduction to fastCNV" 
output: 
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{Introduction to fastCNV}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Context

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE, message = FALSE, warning = FALSE,
  comment = "#>"
)
```

`fastCNV` is a package to detect the putative Copy Number Variations (CNVs) in single cell (scRNAseq) data or Spatial Transcriptomics (ST) data. It also plots the computed CNVs.

In this vignette you will learn how to run this package on single scRNAseq or ST Seurat objects and on a list of Seurat objects. 

## Load packages

To get the example data used in this vignette, you'll need to install and load the `fastCNVdata` package. 
```{r install, eval=FALSE}
remotes::install_github("must-bioinfo/fastCNV")
remotes::install_github("must-bioinfo/fastCNVdata")
```


```{r load}
library(fastCNV)
library(fastCNVdata)
```

## Load the example data

`fastCNV` package works both on scRNAseq data or ST data, and we will demonstrate both cases here. 

Here, we load scColon (scRNAseq data from a colorectal tumor) and crcsample1, crcsample2, crcsample3 and crcsample4 (spatial transcriptomics data from colorectal tumors).

``` {r imports}
utils::data(scColon)
utils::data(crcsample1)
utils::data(crcsample2)
utils::data(crcsample3)
utils::data(crcsample4)
```

## Check annotation

You can load a separate annotation file for your dataset if this information is not available in the Seurat object. In this example, cell type annotations are already present under the column `"annot"` of the metadata.

```{r annots, eval=FALSE}
# Import the annotation file corresponding to your sample. For 10X ST data you can annotate your spots with LoupeBrowser
annotation_file <- read.csv("/path/to/your/annotations/for/crcsample1.csv")

crcsample[["annot"]] <- annotation_file$Annot
```

```{r see_annots}
unique(scColon[["annot"]])
unique(crcsample1[["annot"]])
unique(crcsample2[["annot"]])
unique(crcsample3[["annot"]])
unique(crcsample4[["annot"]])
```


## Run fastCNV 

During this step, fastCNV() will use: 

- `prepareCountsForCNVAnalysis` : runs the Seurat standard clustering algorithm and then aggregates the observations (cells or spots) to into metaspots with up to the number of counts defined by `aggregFactor` (default 30,000). In addition, the observations can be aggregated on their seurat cluster AND their cell type combined by leaving default `aggregateByVar = TRUE` and specifying parameter `referenceVar`.  If the Seurat object has previously been clustered, the clustering will be re-done on sctransformed (SCT) data using 10 PCs with default parameters to `FindNeighbors` and `FindClusters`. This can be skipped by setting `reClusterSeurat = FALSE`.

- `CNVanalysis` : computes the CNV. If you have annotations for your Seurat object, you can add the parameters `referenceVar` and `referenceLabel` to `CNVcalling`, as we are doing for the `crcsample1` example. If given a list of Seurat objects, it will also output a PDF file containing the CNV heatmaps per category accross samples. You can turn this off by setting `doRecapPlot` = `FALSE`.

- `plotCNVResults` : to visualize the results (stored in the assays slot of the Seurat data). By default, the parameter `downsizePlot` is set to `FALSE`, which builds a detailed plot, but takes an important time to render. If desired, you can set the parameter to `TRUE`, which decreases the rendering time by plotting the results at the meta-cell level instead of the cell-level, thus decreasing the definition of the CNV results plotted.  
This function will also build a PDF file for each sample containing their corresponding CNV heatmap in the current working directory, which can be changed using the `savePath` parameter.

- `CNVfractionPerChromosome` : to compute the CNV fraction per chromosome arm. This part of `fastCNV()` can be skipped by turning `getCNVFractionPerChromosome` to `FALSE`.  

We are first going to run `fastCNV()` on our scColon object, taking as reference the cells labeled as `Bcells`, `DC6.pDC`,`EndoArteriel`, `Fibro`, `Mastocytes`, `MyeloLineage`, `NK`, `Pericytes`, `PlasmaCells`, `Schwann.Entericglial`, `SmoothMuscle` and `Tcells`.

```{r run1, fig.width = 8, fig.height=6}
scColon <- fastCNV(seuratObj = scColon, sampleName = "scColon", referenceVar =
                     "annot", referenceLabel = c("Bcells", "DC6.pDC", "EndoArteriel",
                     "Fibro", "Mastocytes", "MyeloLineage", "NK", "Pericytes",
                     "PlasmaCells", "Schwann.Entericglial", "SmoothMuscle","Tcells"))

```

We now run `fastCNV()`on the crcsample1 object We decide to take `epithelium&submucosa`, `submucosa` and `non neo epithelium` as reference since these are non-tumor spots. 

```{r run2, fig.width = 8, fig.height=6}
crcsample1 <- fastCNV(crcsample1, sampleName = "crcsample1", referenceVar = "annot",
                      referenceLabel = c("epithelium&submucosa","submucosa","non neo epithelium"))
```

Last, we are going to run the `fastCNV()` function on a list of Seurat objects. This time, an additional recap plot will be output in a PDF file, containing the CNV heatmaps per category across samples, since there is more than one Seurat object/sample, but you can turn this off by setting `doRecapPlot = FALSE`. 

```{r run3, fig.width = 8, fig.height=6}
seuratList <- c(crcsample2,crcsample3,crcsample4)
sampleNames <- c("crcsample2", "crcsample3", "crcsample4")
names(seuratList) <- sampleNames
referencelabels <- c("epithelium&submucosa","submucosa","non neo epithelium",
                     "squamous epithelium", "epithelium&lam propria", "lamina propria")

seuratList <- fastCNV(seuratList, sampleNames, referenceVar = "annot", 
                      referenceLabel = referencelabels)

```


## CNV fraction

fastCNV also computes a `cnv_fraction` for each observation in the Seurat object. This can be directly plotted using Seurat plotting functions, as we'll demonstrate:

```{r umap, output=FALSE, results='hide'}
library(Seurat)
scColon <- RunUMAP(scColon, dims = 1:10)
```

```{r plot_cnv_umap, fig.width=8, fig.height=4}
FeaturePlot(scColon, features = "cnv_fraction", reduction = "umap", ) |
  DimPlot(scColon, reduction = "umap", group.by =  "annot")
```

Here, we see some clusters with much higher CNV fractions than others. We can directly plot and test this:

```{r plot_cnv_fraction, fig.width=6, fig.height=5}
library(ggplot2)
ggplot(FetchData(scColon, vars = c("annot", "cnv_fraction")), 
       aes(annot, cnv_fraction, fill = annot)) +
  geom_boxplot() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, color = "black"))
```

Of course, we can also plot `cnv_fraction` with the Spatial plotting functions from Seurat, where we can see the non-tumor areas well delimited: 

```{r plot_cnv_fraction2, fig.width=8, fig.height=5}
library(patchwork)
SpatialFeaturePlot(crcsample1, "cnv_fraction", pt.size.factor = 45) | SpatialPlot(crcsample1, group.by = "annot", pt.size.factor = 45)
```

We can also plot the CNV fraction per chromosome arm with the plotting functions from Seurat.

```{r plot_cnv_fraction_2q, fig.width=8, fig.height=5}
SpatialFeaturePlot(crcsample1, features = "2q", pt.size.factor = 45) | SpatialPlot(crcsample1, group.by = "annot", pt.size.factor = 45)

SpatialFeaturePlot(crcsample1, features = "6q", pt.size.factor = 45) | SpatialPlot(crcsample1, group.by = "annot", pt.size.factor = 45)

```

Here we see for example that the right part of the tumor have a high CNV in the chromosome 2q, while the lower and left part of the tumor do not. 
Meanwhile, the opposite happens for chromosome 6q.  

## CNV clusters

In addition to the CNV fraction, we can obtain clusters based on the CNV matrix by running `CNVcluster()`. This runs standard Seurat clustering pipeline on the CNV scores per observation -- uses the first 10 PCs to building the neighborhood graph and then Louvain clustering to find the clusters.

```{r cluster_cnv, results='hide'}
scColon <- CNVcluster(scColon, resolution = 0.8)
crcsample1 <- CNVcluster(crcsample1, resolution = 0.8)
```

Results of `CNVcluster` are saved in the Seurat object under the `"cnv_clusters"` metadata variable, which can be plotted with Seurat functions. We will demonstrate on a ST object, but it also works with regular scRNAseq data.

```{r plot_cnv_clusters, fig.width=6, fig.height=4}
SpatialDimPlot(crcsample1, group.by = "cnv_clusters", pt.size.factor = 45)
```

And we can retrieve this metadata and plot it directly. Here, we relate the `cnv_clusters` with our annotation variable `annot`.

```{r cnv_props, fig.width = 5, fig.height = 4}
library(ggplot2)
library(SeuratObject)
ggplot(FetchData(crcsample1, vars = c("cnv_clusters", "annot")), aes(annot, fill = cnv_clusters)) +
  geom_bar(position = "fill") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, color = "black"))
```

We find that tumor tissue are virtually always cnv_cluster 1, 2 or 3 ; meanwhile, normal labels tend to be cnv_cluster 0, 4, 5, 6 or 7 which present very few or no CNV.


Note that we can also use cnv_clusters to label the CNV heatmap, directly using the parameter `splitPlotOfVar` of the `plotCNVResults()` function.

```{r hm_cnv_cl, fig.width = 7, fig.height=5}
plotCNVResults(scColon, referenceVar = "annot", splitPlotOnVar = "cnv_clusters")
plotCNVResults(crcsample1, referenceVar = "annot", splitPlotOnVar = "cnv_clusters")
```
