<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>fastCNV • fastCNV</title>
<script src="lightswitch.js"></script><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="fastCNV">
<meta name="description" content="Infer CNVs in single-cell or spatial RNA-seq">
<meta property="og:description" content="Infer CNVs in single-cell or spatial RNA-seq">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">fastCNV</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/fastCNV_HD.html">Introduction to fastCNV - Visium HD data</a></li>
    <li><a class="dropdown-item" href="articles/fastCNV_sc.html">Introduction to fastCNV - scRNAseq data</a></li>
    <li><a class="dropdown-item" href="articles/fastCNV_ST.html">Introduction to fastCNV - ST data</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level2">
<h2 id="introduction-">Introduction <img src="https://must-bioinfo.github.io/fastCNV/articles/logo.png" align="right" width="200"><a class="anchor" aria-label="anchor" href="#introduction-"></a>
</h2>
<p><code>fastCNV</code> is an R package that helps you detect, plot and analyse the putative Copy Number Variations (CNVs) in single cell (scRNA-seq) data or Spatial Transcriptomics (ST) data, including Visium HD. Built on <code>SeuratObject</code>, it is easily integrated into scRNA-seq or ST pipelines.</p>
<p><em>WARNING:</em> Project is still under construction and function usage may change.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install fastCNV, run in R:</p>
<pre><code><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"must-bioinfo/fastCNV"</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>Extensive tutorials to run <code>fastCNV</code>on scRNA-seq and Spatial Transcriptomics data are available to get started <a href="https://must-bioinfo.github.io/fastCNV/articles/index.html">here</a>.</p>
</div>
<div class="section level2">
<h2 id="new--run-fastcnv-on-your-visium-hd-data">NEW : RUN fastCNV ON YOUR VISIUM HD DATA<a class="anchor" aria-label="anchor" href="#new--run-fastcnv-on-your-visium-hd-data"></a>
</h2>
<p>You can now run <code>fastCNV</code> on visium HD data using <code><a href="reference/fastCNV_10XHD.html">fastCNV_10XHD()</a></code> <img src="https://must-bioinfo.github.io/fastCNV/articles/fastCNV_HD_files/figure-html/cnvheatmapt_clusters-1.png"></p>
<p><img src="https://must-bioinfo.github.io/fastCNV/articles/fastCNV_HD_files/figure-html/plot_cnv_clusters-1.png" width="45%"><img src="https://must-bioinfo.github.io/fastCNV/articles/fastCNV_HD_files/figure-html/tree_cnv-1.png" width="53%"></p>
</div>
<div class="section level2">
<h2 id="faq">FAQ<a class="anchor" aria-label="anchor" href="#faq"></a>
</h2>
<div class="section level3">
<h3 id="id_1-what-kind-of-data-does-fastcnv-work-on">1. What kind of data does fastCNV work on?<a class="anchor" aria-label="anchor" href="#id_1-what-kind-of-data-does-fastcnv-work-on"></a>
</h3>
<p>Currently, fastCNV can be used on <strong>scRNA-seq</strong>, <strong>Visium</strong>, and <strong>Visium HD</strong> data.</p>
<hr>
</div>
<div class="section level3">
<h3 id="id_2-does-it-work-on-mouse-data">2. Does it work on mouse data?<a class="anchor" aria-label="anchor" href="#id_2-does-it-work-on-mouse-data"></a>
</h3>
<p>Not yet — fastCNV currently only works on <strong>human data</strong>. Support for mouse data is in development.</p>
<hr>
</div>
<div class="section level3">
<h3 id="id_3-how-long-does-it-take-to-run">3. How long does it take to run?<a class="anchor" aria-label="anchor" href="#id_3-how-long-does-it-take-to-run"></a>
</h3>
<p>It depends on the dataset size and computational resources:<br>
- ~ <strong>1 minute</strong> for small scRNA-seq datasets (~4,000 cells)<br>
- ~ <strong>40 minutes</strong> for a Visium HD slide at 16 µm resolution (~150,000 spots)</p>
<hr>
</div>
<div class="section level3">
<h3 id="id_4-i-have-multiple-samples-whats-the-fastest-way-to-run-fastcnv">4. I have multiple samples. What’s the fastest way to run fastCNV?<a class="anchor" aria-label="anchor" href="#id_4-i-have-multiple-samples-whats-the-fastest-way-to-run-fastcnv"></a>
</h3>
<p>You can process all your samples together by providing <code><a href="reference/fastCNV.html">fastCNV()</a></code> with a list of samples, e.g.:</p>
<pre><code><span><span class="fu"><a href="reference/fastCNV.html">fastCNV</a></span><span class="op">(</span>seuratObj <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">sample1</span>, <span class="va">sample2</span>, <span class="va">sample3</span>, <span class="va">...</span><span class="op">)</span>,</span>
<span>        sampleName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span>, <span class="st">"sample3"</span>, <span class="va">...</span><span class="op">)</span>,</span>
<span>        referenceVar <span class="op">=</span> <span class="st">"Annotations"</span>,</span>
<span>        referenceLabel <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Healthy1"</span>, <span class="st">"Healthy2"</span>, <span class="st">"Healthy3"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="id_5-i-have-4-visium-samples-but-only-one-contains-healthy-tissue-can-i-still-run-them-all">5. I have 4 Visium samples, but only one contains healthy tissue. Can I still run them all?<a class="anchor" aria-label="anchor" href="#id_5-i-have-4-visium-samples-but-only-one-contains-healthy-tissue-can-i-still-run-them-all"></a>
</h3>
<p>Yes. When you provide a list of samples, fastCNV will <strong>pool a reference across all of them</strong>, so the healthy tissue from one slide will be used as a reference for all 4 samples.</p>
<hr>
</div>
<div class="section level3">
<h3 id="id_6-can-i-run-fastcnv-without-annotated-samples">6. Can I run fastCNV without annotated samples?<a class="anchor" aria-label="anchor" href="#id_6-can-i-run-fastcnv-without-annotated-samples"></a>
</h3>
<p>Yes, fastCNV can run <strong>without a reference</strong>.<br>
However, we <strong>highly recommend</strong> using a healthy reference. If you don’t have one, you can download a healthy sample from a database (same organ, same technology) and use it as reference.</p>
<hr>
</div>
<div class="section level3">
<h3 id="id_7-what-computational-resources-are-required-for-visium-hd-data">7. What computational resources are required for Visium HD data?<a class="anchor" aria-label="anchor" href="#id_7-what-computational-resources-are-required-for-visium-hd-data"></a>
</h3>
<p>It depends on the bin size:<br>
- <strong>16 µm bin size</strong> → ~64 GB RAM is enough<br>
- <strong>8 µm bin size</strong> → up to ~200 GB RAM may be required</p>
<p>We are working on reducing the resource requirements.</p>
<hr>
</div>
<div class="section level3">
<h3 id="id_8-what-does-the-preparecounts-function-do">8. What does the <code>prepareCounts()</code> function do?<a class="anchor" aria-label="anchor" href="#id_8-what-does-the-preparecounts-function-do"></a>
</h3>
<p>This function is designed for <strong>Visium samples with low read counts</strong>.<br>
It aggregates nearby spots until a threshold is reached (typically 3–4 spots).</p>
<hr>
</div>
<div class="section level3">
<h3 id="id_9-how-is-the-heatmap-generated">9. How is the heatmap generated?<a class="anchor" aria-label="anchor" href="#id_9-how-is-the-heatmap-generated"></a>
</h3>
<p>Heatmaps are produced using <strong>ComplexHeatmap</strong>.<br>
- You can customize annotation colors with <code>referencePalette</code> or <code>clusterPalette</code>.<br>
- You can also choose which annotations to display and how to split the data.</p>
<hr>
</div>
<div class="section level3">
<h3 id="id_10-how-does-fastcnv-run-on-visium-hd-data">10. How does fastCNV run on Visium HD data?<a class="anchor" aria-label="anchor" href="#id_10-how-does-fastcnv-run-on-visium-hd-data"></a>
</h3>
<p>By default, fastCNV uses the <strong>16 µm bin size</strong>. You can increase the resolution, but note that higher resolution requires more computational resources.<br>
Apart from binning and clustering, the pipeline works the same way as for other supported data types.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="capabilities">Capabilities<a class="anchor" aria-label="anchor" href="#capabilities"></a>
</h2>
<p><code>fastCNV</code> can plot a heatmap of inferred CNVs:</p>
<p><img src="https://must-bioinfo.github.io/fastCNV/articles/fastCNV_sc_files/figure-html/heatmap_sc.png"></p>
<p>It also calculates a <code>cnv_fraction</code>, which can be plotted with <code>Seurat</code> standard plotting functions:</p>
<p><img src="https://must-bioinfo.github.io/fastCNV/articles/fastCNV_sc_files/figure-html/plot_cnv_umap-1.png"></p>
<p><code>cnv_fractions</code> can also be visualized spatially for Spatial Transcriptomics samples:</p>
<p><img src="https://must-bioinfo.github.io/fastCNV/articles/fastCNV_ST_files/figure-html/plot_cnv_fraction-1.png"></p>
<p>And <code>cnv_fractions</code> can be used to obtain clonal clusters (<code>cnv_clusters</code>):</p>
<p><img src="https://must-bioinfo.github.io/fastCNV/articles/fastCNV_sc_files/figure-html/plot_cnv_clusters-1.png" alt="Clonal clusters for sc data" width="50%"></p>
<p>It also builds a subclonality tree based on the CNV clusters:</p>
<p><img src="https://must-bioinfo.github.io/fastCNV/articles/fastCNV_sc_files/figure-html/tree_cnv-1.png" alt="Subclonality tree" width="500"></p>
</div>

  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing fastCNV</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Gadea Cabrejas <br><small class="roles"> Author, maintainer </small>  </li>
<li>Clarice Groeneveld <br><small class="roles"> Author, contributor </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gadea Cabrejas, Clarice Groeneveld.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
