<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Introduction to fastCNV • fastCNV</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to fastCNV">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">fastCNV</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.6</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/fastCNV.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Introduction to fastCNV</h1>
            
      

      <div class="d-none name"><code>fastCNV.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="context">Context<a class="anchor" aria-label="anchor" href="#context"></a>
</h2>
<p><code>fastCNV</code> is a package to detect the putative Copy Number
Variations (CNVs) in single cell (scRNAseq) data or Spatial
Transcriptomics (ST) data. It also plots the computed CNVs.</p>
<p>In this vignette you will learn how to run this package on single
scRNAseq or ST Seurat objects and on a list of Seurat objects.</p>
</div>
<div class="section level2">
<h2 id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h2>
<p>To get the example data used in this vignette, you’ll need to install
and load the <code>fastCNVdata</code> package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"must-bioinfo/fastCNV"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"must-bioinfo/fastCNVdata"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">fastCNV</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">fastCNVdata</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-the-example-data">Load the example data<a class="anchor" aria-label="anchor" href="#load-the-example-data"></a>
</h2>
<p><code>fastCNV</code> package works both on scRNAseq data or ST data,
and we will demonstrate both cases here.</p>
<p>Here, we load scColon (scRNAseq data from a colorectal tumor) and
crcsample1, crcsample2, crcsample3 and crcsample4 (spatial
transcriptomics data from colorectal tumors).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">scColon</span><span class="op">)</span></span>
<span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">crcsample1</span><span class="op">)</span></span>
<span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">crcsample2</span><span class="op">)</span></span>
<span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">crcsample3</span><span class="op">)</span></span>
<span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">crcsample4</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="check-annotation">Check annotation<a class="anchor" aria-label="anchor" href="#check-annotation"></a>
</h2>
<p>You can load a separate annotation file for your dataset if this
information is not available in the Seurat object. In this example, cell
type annotations are already present under the column
<code>"annot"</code> of the metadata.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Import the annotation file corresponding to your sample. For 10X ST data you can annotate your spots with LoupeBrowser</span></span>
<span><span class="va">annotation_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"/path/to/your/annotations/for/crcsample1.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">crcsample</span><span class="op">[[</span><span class="st">"annot"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">annotation_file</span><span class="op">$</span><span class="va">Annot</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">scColon</span><span class="op">[[</span><span class="st">"annot"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                                       annot</span></span>
<span><span class="co">#&gt; C113_N_1_1_0_c1_v2_id-AAACCTGTCAGTGTTG               normal</span></span>
<span><span class="co">#&gt; C113_N_1_1_0_c1_v2_id-AAAGCAACACCCTATC                Fibro</span></span>
<span><span class="co">#&gt; C113_N_1_1_0_c1_v2_id-AAATGCCGTCACTTCC                   NK</span></span>
<span><span class="co">#&gt; C113_N_1_1_0_c1_v2_id-AACGTTGAGAGGGATA          PlasmaCells</span></span>
<span><span class="co">#&gt; C113_N_1_1_0_c1_v2_id-AACTCTTGTCTCACCT         MyeloLineage</span></span>
<span><span class="co">#&gt; C113_N_1_1_0_c1_v2_id-ACCCACTTCGAATGCT               Tcells</span></span>
<span><span class="co">#&gt; C113_N_1_1_0_c1_v2_id-ACTGCTCAGACACTAA               Bcells</span></span>
<span><span class="co">#&gt; C113_N_1_1_0_c1_v2_id-CGGAGTCTCACCAGGC         SmoothMuscle</span></span>
<span><span class="co">#&gt; C113_N_1_1_0_c1_v2_id-CGGGTCATCTACCAGA           Mastocytes</span></span>
<span><span class="co">#&gt; C113_N_1_1_0_c1_v2_id-GTTAAGCTCACAGTAC         EndoArteriel</span></span>
<span><span class="co">#&gt; C113_T_1_1_0_c1_v2_id-AAACCTGAGATCGATA                 EpiT</span></span>
<span><span class="co">#&gt; C113_T_1_1_0_c1_v2_id-ATCATGGGTCGTCTTC              DC6.pDC</span></span>
<span><span class="co">#&gt; C113_T_1_1_0_c1_v2_id-CGATTGATCTGCGACG            Pericytes</span></span>
<span><span class="co">#&gt; C113_T_1_1_1_c1_v2_id-AGCGGTCAGGCTCAGA Schwann.Entericglial</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">crcsample1</span><span class="op">[[</span><span class="st">"annot"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                          annot</span></span>
<span><span class="co">#&gt; AAACAAGTATCTCCCA-1        epithelium&amp;submucosa</span></span>
<span><span class="co">#&gt; AAACACCAATAACTGC-1                     exclude</span></span>
<span><span class="co">#&gt; AAACATTTCCCGGATT-1                   submucosa</span></span>
<span><span class="co">#&gt; AAACCCGAACGAAATC-1                       tumor</span></span>
<span><span class="co">#&gt; AAACCTAAGCAGCCGG-1 tumor&amp;stroma_IC med to high</span></span>
<span><span class="co">#&gt; AAAGGCTACGGACCAT-1 stroma_fibroblastic_IC high</span></span>
<span><span class="co">#&gt; AAAGTTGACTCCCGTA-1          non neo epithelium</span></span>
<span><span class="co">#&gt; AAATAGGGTGCTATTG-1      IC aggregate_submucosa</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">crcsample2</span><span class="op">[[</span><span class="st">"annot"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                          annot</span></span>
<span><span class="co">#&gt; AAACAAGTATCTCCCA-1 tumor&amp;stroma_IC med to high</span></span>
<span><span class="co">#&gt; AAACAGCTTTCAGAAG-1                       tumor</span></span>
<span><span class="co">#&gt; AAACAGGGTCTATATT-1 stroma_fibroblastic_IC high</span></span>
<span><span class="co">#&gt; AAACCGTTCGTCCAGG-1  stroma_fibroblastic_IC med</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">crcsample3</span><span class="op">[[</span><span class="st">"annot"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                                      annot</span></span>
<span><span class="co">#&gt; AAACAGAGCGACTCCT-1                          lamina propria</span></span>
<span><span class="co">#&gt; AAACCGGGTAGGTACC-1 connective tissue_3_fibroblastic_IC med</span></span>
<span><span class="co">#&gt; AAACCGTTCGTCCAGG-1 connective tissue_2_fibroblastic_IC low</span></span>
<span><span class="co">#&gt; AAACCTCATGAAGTTG-1        connective tissue_6_hemosiderin?</span></span>
<span><span class="co">#&gt; AAACGAGACGGTTGAT-1                  IC aggregate_submucosa</span></span>
<span><span class="co">#&gt; AAACTGCTGGCTCCAA-1               connective tissue_1_edema</span></span>
<span><span class="co">#&gt; AAATAACCATACGGGA-1                  epithelium&amp;lam propria</span></span>
<span><span class="co">#&gt; AAATAGGGTGCTATTG-1                     squamous epithelium</span></span>
<span><span class="co">#&gt; AAATGGCATGTCTTGT-1                                 exclude</span></span>
<span><span class="co">#&gt; AACCTTTACGACGTCT-1                      non neo epithelium</span></span>
<span><span class="co">#&gt; AACGTGCGAAAGTCTC-1   connective tissue_4_muscularis_IC low</span></span>
<span><span class="co">#&gt; ACAAGGGCAGGCTCTG-1                        glandular tissue</span></span>
<span><span class="co">#&gt; ACGTTAGATTTGCCCG-1          IC aggregate connective tissue</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">crcsample4</span><span class="op">[[</span><span class="st">"annot"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                                 annot</span></span>
<span><span class="co">#&gt; AAACAATCTACTAGCA-1        stroma_fibroblastic_IC high</span></span>
<span><span class="co">#&gt; AAACAGAGCGACTCCT-1        tumor&amp;stroma_IC med to high</span></span>
<span><span class="co">#&gt; AAACGAAGAACATACC-1                              tumor</span></span>
<span><span class="co">#&gt; AAAGGGCAGCTTGAAT-1         stroma desmoplastic_IC low</span></span>
<span><span class="co">#&gt; AAAGTGTGATTTATCT-1                          submucosa</span></span>
<span><span class="co">#&gt; AAAGTTGACTCCCGTA-1                 non neo epithelium</span></span>
<span><span class="co">#&gt; AAATACCTATAAGCAT-1          muscularis_IC med to high</span></span>
<span><span class="co">#&gt; AAATGTGGGTGCTCCT-1         stroma_fibroblastic_IC med</span></span>
<span><span class="co">#&gt; AACAATTACTCTACGC-1                            exclude</span></span>
<span><span class="co">#&gt; AACTCTCAATAGAGCG-1               epithelium&amp;submucosa</span></span>
<span><span class="co">#&gt; AAGACTAACCCGTTGT-1 stroma desmoplastic_IC med to high</span></span>
<span><span class="co">#&gt; AAGCTAGATCGAGTAA-1             IC aggregate_submucosa</span></span>
<span><span class="co">#&gt; ACCGATGGTAGCATCG-1  IC aggregate_stroma or muscularis</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="run-fastcnv">Run fastCNV<a class="anchor" aria-label="anchor" href="#run-fastcnv"></a>
</h2>
<p>During this step, fastCNV() will use:</p>
<ul>
<li><p><code>prepareCountsForCNVAnalysis</code> : runs the Seurat
standard clustering algorithm and then aggregates the observations
(cells or spots) to into metaspots with up to the number of counts
defined by <code>aggregFactor</code> (default 30,000). In addition, the
observations can be aggregated on their seurat cluster AND their cell
type combined by leaving default <code>aggregateByVar = TRUE</code> and
specifying parameter <code>referenceVar</code>. If the Seurat object has
previously been clustered, the clustering will be re-done on
sctransformed (SCT) data using 10 PCs with default parameters to
<code>FindNeighbors</code> and <code>FindClusters</code>. This can be
skipped by setting <code>reClusterSeurat = FALSE</code>.</p></li>
<li><p><code>CNVanalysis</code> : computes the CNV. If you have
annotations for your Seurat object, you can add the parameters
<code>referenceVar</code> and <code>referenceLabel</code> to
<code>CNVcalling</code>, as we are doing for the <code>crcsample1</code>
example. If given a list of Seurat objects, it will also output a PDF
file containing the CNV heatmaps per category accross samples. You can
turn this off by setting <code>doRecapPlot</code> =
<code>FALSE</code>.</p></li>
<li><p><code>plotCNVResults</code> : to visualize the results (stored in
the assays slot of the Seurat data). By default, the parameter
<code>downsizePlot</code> is set to <code>FALSE</code>, which builds a
detailed plot, but takes an important time to render. If desired, you
can set the parameter to <code>TRUE</code>, which decreases the
rendering time by plotting the results at the meta-cell level instead of
the cell-level, thus decreasing the definition of the CNV results
plotted.<br>
This function will also build a PDF file for each sample containing
their corresponding CNV heatmap in the current working directory, which
can be changed using the <code>savePath</code> parameter.</p></li>
<li><p><code>CNVfractionPerChromosome</code> : to compute the CNV
fraction per chromosome arm. This part of <code><a href="../reference/fastCNV.html">fastCNV()</a></code> can be
skipped by turning <code>getCNVFractionPerChromosome</code> to
<code>FALSE</code>.</p></li>
</ul>
<p>We are first going to run <code><a href="../reference/fastCNV.html">fastCNV()</a></code> on our scColon
object, taking as reference the cells labeled as <code>Bcells</code>,
<code>DC6.pDC</code>,<code>EndoArteriel</code>, <code>Fibro</code>,
<code>Mastocytes</code>, <code>MyeloLineage</code>, <code>NK</code>,
<code>Pericytes</code>, <code>PlasmaCells</code>,
<code>Schwann.Entericglial</code>, <code>SmoothMuscle</code> and
<code>Tcells</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scColon</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fastCNV.html">fastCNV</a></span><span class="op">(</span>seuratObj <span class="op">=</span> <span class="va">scColon</span>, sampleName <span class="op">=</span> <span class="st">"scColon"</span>, referenceVar <span class="op">=</span></span>
<span>                     <span class="st">"annot"</span>, referenceLabel <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Bcells"</span>, <span class="st">"DC6.pDC"</span>, <span class="st">"EndoArteriel"</span>,</span>
<span>                     <span class="st">"Fibro"</span>, <span class="st">"Mastocytes"</span>, <span class="st">"MyeloLineage"</span>, <span class="st">"NK"</span>, <span class="st">"Pericytes"</span>,</span>
<span>                     <span class="st">"PlasmaCells"</span>, <span class="st">"Schwann.Entericglial"</span>, <span class="st">"SmoothMuscle"</span>,<span class="st">"Tcells"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Aggregating counts matrix."</span></span>
<span><span class="co">#&gt; [1] "Running Seurat SCTransform and clustering. This could take some time."</span></span>
<span><span class="co">#&gt; [1] "Sample : scColon"</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 2840</span></span>
<span><span class="co">#&gt; Number of edges: 83853</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.8916</span></span>
<span><span class="co">#&gt; Number of communities: 15</span></span>
<span><span class="co">#&gt; Elapsed time: 0 seconds</span></span>
<span><span class="co">#&gt; [1] "Seurat SCTransform and clustering done for sample scColon."</span></span>
<span><span class="co">#&gt; [1] "Running CNVAnalysis"</span></span>
<span><span class="co">#&gt; [1] "CNVAnalysis done!"</span></span>
<span><span class="co">#&gt; [1] "Plotting CNV results. This step may take some time."</span></span></code></pre></div>
<p><img src="fastCNV_files/figure-html/run1-1.png" width="768"></p>
<p>We now run <code><a href="../reference/fastCNV.html">fastCNV()</a></code>on the crcsample1 object We decide
to take <code>epithelium&amp;submucosa</code>, <code>submucosa</code>
and <code>non neo epithelium</code> as reference since these are
non-tumor spots.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">crcsample1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fastCNV.html">fastCNV</a></span><span class="op">(</span><span class="va">crcsample1</span>, sampleName <span class="op">=</span> <span class="st">"crcsample1"</span>, referenceVar <span class="op">=</span> <span class="st">"annot"</span>,</span>
<span>                      referenceLabel <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"epithelium&amp;submucosa"</span>,<span class="st">"submucosa"</span>,<span class="st">"non neo epithelium"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Aggregating counts matrix."</span></span>
<span><span class="co">#&gt; [1] "Running Seurat SCTransform and clustering. This could take some time."</span></span>
<span><span class="co">#&gt; [1] "Sample : crcsample1"</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 2203</span></span>
<span><span class="co">#&gt; Number of edges: 65483</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.8181</span></span>
<span><span class="co">#&gt; Number of communities: 10</span></span>
<span><span class="co">#&gt; Elapsed time: 0 seconds</span></span>
<span><span class="co">#&gt; [1] "Seurat SCTransform and clustering done for sample crcsample1."</span></span>
<span><span class="co">#&gt; [1] "Running CNVAnalysis"</span></span>
<span><span class="co">#&gt; [1] "CNVAnalysis done!"</span></span>
<span><span class="co">#&gt; [1] "Plotting CNV results. This step may take some time."</span></span></code></pre></div>
<p><img src="fastCNV_files/figure-html/run2-1.png" width="768"></p>
<p>Last, we are going to run the <code><a href="../reference/fastCNV.html">fastCNV()</a></code> function on a
list of Seurat objects. This time, an additional recap plot will be
output in a PDF file, containing the CNV heatmaps per category across
samples, since there is more than one Seurat object/sample, but you can
turn this off by setting <code>doRecapPlot = FALSE</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seuratList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">crcsample2</span>,<span class="va">crcsample3</span>,<span class="va">crcsample4</span><span class="op">)</span></span>
<span><span class="va">sampleNames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"crcsample2"</span>, <span class="st">"crcsample3"</span>, <span class="st">"crcsample4"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">seuratList</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sampleNames</span></span>
<span><span class="va">referencelabels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"epithelium&amp;submucosa"</span>,<span class="st">"submucosa"</span>,<span class="st">"non neo epithelium"</span>,</span>
<span>                     <span class="st">"squamous epithelium"</span>, <span class="st">"epithelium&amp;lam propria"</span>, <span class="st">"lamina propria"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seuratList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fastCNV.html">fastCNV</a></span><span class="op">(</span><span class="va">seuratList</span>, <span class="va">sampleNames</span>, referenceVar <span class="op">=</span> <span class="st">"annot"</span>, </span>
<span>                      referenceLabel <span class="op">=</span> <span class="va">referencelabels</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Aggregating counts matrix."</span></span>
<span><span class="co">#&gt; [1] "Running Seurat SCTransform and clustering. This could take some time."</span></span>
<span><span class="co">#&gt; [1] "Sample : crcsample2"</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 2317</span></span>
<span><span class="co">#&gt; Number of edges: 69983</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.8131</span></span>
<span><span class="co">#&gt; Number of communities: 15</span></span>
<span><span class="co">#&gt; Elapsed time: 0 seconds</span></span>
<span><span class="co">#&gt; [1] "Seurat SCTransform and clustering done for sample crcsample2."</span></span>
<span><span class="co">#&gt; [1] "Running Seurat SCTransform and clustering. This could take some time."</span></span>
<span><span class="co">#&gt; [1] "Sample : crcsample3"</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 1656</span></span>
<span><span class="co">#&gt; Number of edges: 46705</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.8145</span></span>
<span><span class="co">#&gt; Number of communities: 12</span></span>
<span><span class="co">#&gt; Elapsed time: 0 seconds</span></span>
<span><span class="co">#&gt; [1] "Seurat SCTransform and clustering done for sample crcsample3."</span></span>
<span><span class="co">#&gt; [1] "Running Seurat SCTransform and clustering. This could take some time."</span></span>
<span><span class="co">#&gt; [1] "Sample : crcsample4"</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 1691</span></span>
<span><span class="co">#&gt; Number of edges: 48501</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.8099</span></span>
<span><span class="co">#&gt; Number of communities: 9</span></span>
<span><span class="co">#&gt; Elapsed time: 0 seconds</span></span>
<span><span class="co">#&gt; [1] "Seurat SCTransform and clustering done for sample crcsample4."</span></span>
<span><span class="co">#&gt; [1] "Running CNVAnalysis"</span></span>
<span><span class="co">#&gt; [1] "Plotting the CNV recap heatmap per category accross samples. This could take some time."</span></span>
<span><span class="co">#&gt; [1] "CNVAnalysis done!"</span></span>
<span><span class="co">#&gt; [1] "Plotting CNV results. This step may take some time."</span></span></code></pre></div>
<p><img src="fastCNV_files/figure-html/run3-1.png" width="768"><img src="fastCNV_files/figure-html/run3-2.png" width="768"><img src="fastCNV_files/figure-html/run3-3.png" width="768"></p>
</div>
<div class="section level2">
<h2 id="cnv-fraction">CNV fraction<a class="anchor" aria-label="anchor" href="#cnv-fraction"></a>
</h2>
<p>fastCNV also computes a <code>cnv_fraction</code> for each
observation in the Seurat object. This can be directly plotted using
Seurat plotting functions, as we’ll demonstrate:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">scColon</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">scColon</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">scColon</span>, features <span class="op">=</span> <span class="st">"cnv_fraction"</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, <span class="op">)</span> <span class="op">|</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">scColon</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span>  <span class="st">"annot"</span><span class="op">)</span></span></code></pre></div>
<p><img src="fastCNV_files/figure-html/plot_cnv_umap-1.png" width="768"></p>
<p>Here, we see some clusters with much higher CNV fractions than
others. We can directly plot and test this:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/FetchData.html" class="external-link">FetchData</a></span><span class="op">(</span><span class="va">scColon</span>, vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"annot"</span>, <span class="st">"cnv_fraction"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">annot</span>, <span class="va">cnv_fraction</span>, fill <span class="op">=</span> <span class="va">annot</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust <span class="op">=</span> <span class="fl">1</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="fastCNV_files/figure-html/plot_cnv_fraction-1.png" width="576"></p>
<p>Of course, we can also plot <code>cnv_fraction</code> with the
Spatial plotting functions from Seurat, where we can see the non-tumor
areas well delimited:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/SpatialPlot.html" class="external-link">SpatialFeaturePlot</a></span><span class="op">(</span><span class="va">crcsample1</span>, <span class="st">"cnv_fraction"</span>, pt.size.factor <span class="op">=</span> <span class="fl">45</span><span class="op">)</span> <span class="op">|</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SpatialPlot.html" class="external-link">SpatialPlot</a></span><span class="op">(</span><span class="va">crcsample1</span>, group.by <span class="op">=</span> <span class="st">"annot"</span>, pt.size.factor <span class="op">=</span> <span class="fl">45</span><span class="op">)</span></span></code></pre></div>
<p><img src="fastCNV_files/figure-html/plot_cnv_fraction2-1.png" width="768"></p>
<p>We can also plot the CNV per chromosome arm with the plotting
functions from Seurat.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org" class="external-link">scales</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/SpatialPlot.html" class="external-link">SpatialFeaturePlot</a></span><span class="op">(</span><span class="va">crcsample1</span>, features <span class="op">=</span> <span class="st">"11q_CNV"</span>, pt.size.factor <span class="op">=</span> <span class="fl">45</span><span class="op">)</span>  <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_distiller</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"RdBu"</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>                       rescaler <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">to</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, <span class="va">from</span> <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">{</span></span>
<span>                         <span class="fu"><a href="https://scales.r-lib.org/reference/rescale_mid.html" class="external-link">rescale_mid</a></span><span class="op">(</span><span class="va">x</span>, to <span class="op">=</span> <span class="va">to</span>, mid <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span>                       <span class="op">}</span><span class="op">)</span> <span class="op">|</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/SpatialPlot.html" class="external-link">SpatialFeaturePlot</a></span><span class="op">(</span><span class="va">crcsample1</span>, features <span class="op">=</span> <span class="st">"6q_CNV"</span>, pt.size.factor <span class="op">=</span> <span class="fl">45</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_distiller</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"RdBu"</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>                       rescaler <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">to</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, <span class="va">from</span> <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">{</span></span>
<span>                         <span class="fu"><a href="https://scales.r-lib.org/reference/rescale_mid.html" class="external-link">rescale_mid</a></span><span class="op">(</span><span class="va">x</span>, to <span class="op">=</span> <span class="va">to</span>, mid <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span>                       <span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p><img src="fastCNV_files/figure-html/plot_cnv_fraction_2q-1.png" width="768"></p>
<p>Here we see for example that the right part of the tumor have a CNV
loss in the chromosome 11q, while the lower and left part of the tumor
do not. Meanwhile, the opposite happens for chromosome 6q, having a gain
on the lower and left part of the tumor.</p>
</div>
<div class="section level2">
<h2 id="cnv-classification">CNV classification<a class="anchor" aria-label="anchor" href="#cnv-classification"></a>
</h2>
<p>Using the CNV per chromosome arm and <code>CNVclassify()</code> we
can get the alterations per chromosome arm (gain, loss or none).
<code>CNVclassify()</code> takes as arguments a Seurat object containing
in the metadata the CNV per chromosome arm. It calculates 3 peaks of the
CNV, corresponding to a loss, a gain and no alteration. Using these 3
peaks, it builds a table for each cell/spot and each chromosome arm
containing either “loss”, “gain” or “no_alteration”.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">crcsample1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CNVclassification.html">CNVclassification</a></span><span class="op">(</span><span class="va">crcsample1</span>, tumorLabels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tumor"</span>, <span class="st">"tumor&amp;stroma_IC med to high"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/SpatialPlot.html" class="external-link">SpatialDimPlot</a></span><span class="op">(</span><span class="va">crcsample1</span>, group.by <span class="op">=</span> <span class="st">"11p_CNV_classification"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>gain <span class="op">=</span> <span class="st">"red"</span>, no_alteration <span class="op">=</span> <span class="st">"grey"</span>, loss <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span><span class="op">|</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SpatialPlot.html" class="external-link">SpatialDimPlot</a></span><span class="op">(</span><span class="va">crcsample1</span>, group.by <span class="op">=</span> <span class="st">"13q_CNV_classification"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>gain <span class="op">=</span> <span class="st">"red"</span>, no_alteration <span class="op">=</span> <span class="st">"grey"</span>, loss <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="fastCNV_files/figure-html/cnv%20classification-1.png" width="768"></p>
</div>
<div class="section level2">
<h2 id="cnv-clusters">CNV clusters<a class="anchor" aria-label="anchor" href="#cnv-clusters"></a>
</h2>
<p>In addition to the CNV fraction, we can obtain clusters based on the
CNV matrix by running <code><a href="../reference/CNVcluster.html">CNVcluster()</a></code>. This runs standard
Seurat clustering pipeline on the CNV scores per observation – uses the
first 10 PCs to building the neighborhood graph and then Louvain
clustering to find the clusters.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scColon</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CNVcluster.html">CNVcluster</a></span><span class="op">(</span><span class="va">scColon</span>, resolution <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></span>
<span><span class="va">crcsample1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CNVcluster.html">CNVcluster</a></span><span class="op">(</span><span class="va">crcsample1</span>, resolution <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></span></code></pre></div>
<p>Results of <code>CNVcluster</code> are saved in the Seurat object
under the <code>"cnv_clusters"</code> metadata variable, which can be
plotted with Seurat functions. We will demonstrate on a ST object, but
it also works with regular scRNAseq data.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/SpatialPlot.html" class="external-link">SpatialDimPlot</a></span><span class="op">(</span><span class="va">crcsample1</span>, group.by <span class="op">=</span> <span class="st">"cnv_clusters"</span>, pt.size.factor <span class="op">=</span> <span class="fl">45</span><span class="op">)</span></span></code></pre></div>
<p><img src="fastCNV_files/figure-html/plot_cnv_clusters-1.png" width="576"></p>
<p>And we can retrieve this metadata and plot it directly. Here, we
relate the <code>cnv_clusters</code> with our annotation variable
<code>annot</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.github.io/seurat-object/" class="external-link">SeuratObject</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/FetchData.html" class="external-link">FetchData</a></span><span class="op">(</span><span class="va">crcsample1</span>, vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cnv_clusters"</span>, <span class="st">"annot"</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">annot</span>, fill <span class="op">=</span> <span class="va">cnv_clusters</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>position <span class="op">=</span> <span class="st">"fill"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust <span class="op">=</span> <span class="fl">1</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="fastCNV_files/figure-html/cnv_props-1.png" width="480"></p>
<p>We find that tumor tissue are virtually always cnv_cluster 1, 2 or 3
; meanwhile, normal labels tend to be cnv_cluster 0, 4, 5, 6 or 7 which
present very few or no CNV.</p>
<p>Note that we can also use cnv_clusters to label the CNV heatmap,
directly using the parameter <code>splitPlotOfVar</code> of the
<code><a href="../reference/plotCNVResults.html">plotCNVResults()</a></code> function.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCNVResults.html">plotCNVResults</a></span><span class="op">(</span><span class="va">scColon</span>, referenceVar <span class="op">=</span> <span class="st">"annot"</span>, splitPlotOnVar <span class="op">=</span> <span class="st">"cnv_clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="fastCNV_files/figure-html/hm_cnv_cl-1.png" width="672"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCNVResults.html">plotCNVResults</a></span><span class="op">(</span><span class="va">crcsample1</span>, referenceVar <span class="op">=</span> <span class="st">"annot"</span>, splitPlotOnVar <span class="op">=</span> <span class="st">"cnv_clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="fastCNV_files/figure-html/hm_cnv_cl-2.png" width="672"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gadea Cabrejas.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
