<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>fastCNV calls all of the internal functions needed to compute the putative CNV on a seurat object or a list of seurat objects — fastCNV • fastCNV</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="fastCNV calls all of the internal functions needed to compute the putative CNV on a seurat object or a list of seurat objects — fastCNV"><meta name="description" content="fastCNV calls all of the internal functions needed to compute the putative CNV on a seurat object or a list of seurat objects"><meta property="og:description" content="fastCNV calls all of the internal functions needed to compute the putative CNV on a seurat object or a list of seurat objects"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">fastCNV</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.6</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/fastCNV.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>fastCNV calls all of the internal functions needed to compute the putative CNV on a seurat object or a list of seurat objects</h1>

      <div class="d-none name"><code>fastCNV.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>fastCNV calls all of the internal functions needed to compute the putative CNV on a seurat object or a list of seurat objects</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fastCNV</span><span class="op">(</span></span>
<span>  <span class="va">seuratObj</span>,</span>
<span>  <span class="va">sampleName</span>,</span>
<span>  referenceVar <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  referenceLabel <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  splitPlotOnVar <span class="op">=</span> <span class="va">referenceVar</span>,</span>
<span>  downsizePlot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  doPlot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  doRecapPlot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  prepareCounts <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  pooledReference <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  assay <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  getCNVPerChromosomeArm <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  savePath <span class="op">=</span> <span class="st">"."</span>,</span>
<span>  aggregFactor <span class="op">=</span> <span class="fl">15000</span>,</span>
<span>  seuratClusterResolution <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  aggregateByVar <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  reClusterSeurat <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  scaleOnReferenceLabel <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  thresholdPercentile <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  genes <span class="op">=</span> <span class="fu"><a href="getGenes.html">getGenes</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  windowSize <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  windowStep <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  topNGenes <span class="op">=</span> <span class="fl">7000</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-seuratobj">seuratObj<a class="anchor" aria-label="anchor" href="#arg-seuratobj"></a></dt>
<dd><p>Seurat object or list of seurat objects</p></dd>


<dt id="arg-samplename">sampleName<a class="anchor" aria-label="anchor" href="#arg-samplename"></a></dt>
<dd><p>Name of the sample or list of names corresponding to the samples in the <code>seuratObj</code></p></dd>


<dt id="arg-referencevar">referenceVar<a class="anchor" aria-label="anchor" href="#arg-referencevar"></a></dt>
<dd><p>The variable name of the annotations in the Seurat metadata</p></dd>


<dt id="arg-referencelabel">referenceLabel<a class="anchor" aria-label="anchor" href="#arg-referencelabel"></a></dt>
<dd><p>The label given to the observations wanted as reference (can be any type of annotation)</p></dd>


<dt id="arg-splitplotonvar">splitPlotOnVar<a class="anchor" aria-label="anchor" href="#arg-splitplotonvar"></a></dt>
<dd><p>The name given to the meta.data slot to split the observations on during the <code>plotCNVResults</code> step, if different from <code>referenceVar</code>.</p></dd>


<dt id="arg-downsizeplot">downsizePlot<a class="anchor" aria-label="anchor" href="#arg-downsizeplot"></a></dt>
<dd><p>Subset the observations to speed up the plotting process (default = <code>FALSE</code>).</p></dd>


<dt id="arg-doplot">doPlot<a class="anchor" aria-label="anchor" href="#arg-doplot"></a></dt>
<dd><p>If <code>TRUE</code> will plot a heatmap for each of the samples (default = <code>TRUE</code>)</p></dd>


<dt id="arg-dorecapplot">doRecapPlot<a class="anchor" aria-label="anchor" href="#arg-dorecapplot"></a></dt>
<dd><p>Default <code>TRUE</code>. Will output the CNV heatmaps by annotation if <code>TRUE</code>.</p></dd>


<dt id="arg-preparecounts">prepareCounts<a class="anchor" aria-label="anchor" href="#arg-preparecounts"></a></dt>
<dd><p>If <code>FALSE</code> will not run the <code>prepareCountsForCNVAnalysis</code> function. Default <code>TRUE</code>.</p></dd>


<dt id="arg-pooledreference">pooledReference<a class="anchor" aria-label="anchor" href="#arg-pooledreference"></a></dt>
<dd><p>Default <code>TRUE</code>. Will build a pooled reference across all samples if <code>TRUE</code>.</p></dd>


<dt id="arg-assay">assay<a class="anchor" aria-label="anchor" href="#arg-assay"></a></dt>
<dd><p>Name of the assay to run the CNV on. Takes the results of <code>prepareCountsForCNVAnalysis</code> by default if available</p></dd>


<dt id="arg-getcnvperchromosomearm">getCNVPerChromosomeArm<a class="anchor" aria-label="anchor" href="#arg-getcnvperchromosomearm"></a></dt>
<dd><p>If <code>TRUE</code>, will save into the metadata the CNV per chromosome arm</p></dd>


<dt id="arg-savepath">savePath<a class="anchor" aria-label="anchor" href="#arg-savepath"></a></dt>
<dd><p>Default <code>.</code>. If <code>NULL</code> the heatmap won't be saved as a pdf.</p></dd>


<dt id="arg-aggregfactor">aggregFactor<a class="anchor" aria-label="anchor" href="#arg-aggregfactor"></a></dt>
<dd><p>The number of counts per spot desired (default <code>30 000</code>). If less than <code>1000</code>, will not run the <code>prepareCountsForCNVAnalysis</code> function.</p></dd>


<dt id="arg-seuratclusterresolution">seuratClusterResolution<a class="anchor" aria-label="anchor" href="#arg-seuratclusterresolution"></a></dt>
<dd><p>The resolution wanted for the seurat clusters (default 0.8)</p></dd>


<dt id="arg-aggregatebyvar">aggregateByVar<a class="anchor" aria-label="anchor" href="#arg-aggregatebyvar"></a></dt>
<dd><p>If <code>referenceVar</code> is given, whether to use it to pool the observations</p></dd>


<dt id="arg-reclusterseurat">reClusterSeurat<a class="anchor" aria-label="anchor" href="#arg-reclusterseurat"></a></dt>
<dd><p>Whether to re-cluster if the Seurat object given already has a <code>seurat_clusters</code> slot in its meta.data</p></dd>


<dt id="arg-scaleonreferencelabel">scaleOnReferenceLabel<a class="anchor" aria-label="anchor" href="#arg-scaleonreferencelabel"></a></dt>
<dd><p>If you want to scale the results depending on the normal observations</p></dd>


<dt id="arg-thresholdpercentile">thresholdPercentile<a class="anchor" aria-label="anchor" href="#arg-thresholdpercentile"></a></dt>
<dd><p>Which quantiles to take (if 0.01 it will take 0.01-0.99). Background noise appears with higher numbers.</p></dd>


<dt id="arg-genes">genes<a class="anchor" aria-label="anchor" href="#arg-genes"></a></dt>
<dd><p>List of genes (default genes from ensembl version 111)</p></dd>


<dt id="arg-windowsize">windowSize<a class="anchor" aria-label="anchor" href="#arg-windowsize"></a></dt>
<dd><p>Size of the genomic windows</p></dd>


<dt id="arg-windowstep">windowStep<a class="anchor" aria-label="anchor" href="#arg-windowstep"></a></dt>
<dd><p>Step between the genomic windows</p></dd>


<dt id="arg-topngenes">topNGenes<a class="anchor" aria-label="anchor" href="#arg-topngenes"></a></dt>
<dd><p>Number of top expressed genes to keep</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>This function returns a list of the seurat objects after all the analysis and it creates the heatmaps of the CNVs of every object in seuratObj and saves them in a pdf in the current working directory</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gadea Cabrejas.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

